src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt

```diff
@@ -244,8 +244,8 @@ object RustParserUtil : GeneratedParserUtilBase() {
         val bound = enter_section_(b)
         val traitRef = enter_section_(b)
 
-        if (!pathP.parse(b, level + 1)) {
-            // May be it is lifetime `'a` or `for<'a>`
+        if (!pathP.parse(b, level + 1) || nextTokenIs(b, EXCL)) {
```

Comment:
Reject `!` after a path. Needed for changes in `TypeReferenceInner` rule

Reply:
Maybe it makes sense to swap these conditions because `nextTokenIs(b, EXCL)` should be much faster?

Reply:
Order is important b/c `pathP.parse` is not pure (moves a lexer position to the end of a path, if true returned)